src/protocol/yaml.rs

```diff
@@ -3,32 +3,40 @@ use linked_hash_map::LinkedHashMap;
 use std::fmt;
 use std::io;
 use std::io::Cursor;
-use yaml_rust::{yaml::Hash, Yaml, YamlEmitter};
+use yaml_rust::{yaml::HashNode, Node, Yaml, YamlEmitter, YamlMarked, YamlNode};
 
 pub trait YamlExt {
+    type Child;
+
     fn expect_str(&self) -> R<&str>;
 
-    fn expect_array(&self) -> R<&Vec<Yaml>>;
+    fn expect_array(&self) -> R<&Vec<Self::Child>>;
 
-    fn expect_object(&self) -> R<&LinkedHashMap<Yaml, Yaml>>;
+    fn expect_object(&self) -> R<&LinkedHashMap<Self::Child, Self::Child>>;
 
     fn expect_integer(&self) -> R<i32>;
 }
 
-impl YamlExt for Yaml {
+impl<T> YamlExt for T
+where
+    T: fmt::Debug + YamlNode,
+    <T as YamlNode>::Child: fmt::Debug,
```

Comment:
can you help me understand how to interpret this? 

is `<T as YamlNode>::Child` just the way to specify the generic `Child` inside of `T`? then, since we specify on the line above that `T: YamlNode`, this allows us to add bounds to the `Child`?

is there a name for this pattern that i could look up?

Reply:
Yes, you've got it. This is a constraint on an associated type. The methods in `YamlExt` use `Debug` implementations to return errors, which means that both `T` and the `Child` associated type must implement `Debug`. And yes, `<T as YamlNode>::Child` is a way to reference the associated type in the `YamlNode` impl for whatever type `T` happens to be. I don't know the specific term for casting a type variable to a trait to access its associated types, I only know how to describe it in those terms.

Another way to make this work would have been to put the `Debug` constraint on the associated type in the `YamlExt` trait, instead of in the impl like this:

```rust
trait YamlExt {
    type Child: fmt::Debug;
    // ...
}
```

But it is in principle possible to implement `YamlExt` with a `Child` type that does not implement `Debug`; and my instruction in Haskell was to keep trait / type class bounds as close as possible to implementations that actually require them to keep reusable code as general as is reasonable.

Reply:
cool, thanks for the explanation!